### ISSUE TITLE
Investigate intermittent test failure in dropwizard-client-poller

### ISSUE THREAD

**Issue by sleberknight** [kiwiproject/dropwizard-client-poller#131]
Recently we saw this test failure:

```
 [ERROR]   DefaultClientPollerStatisticsTest.testIncrementCount_SetsLastAttemptTime(SoftAssertions) 
    Multiple Failures (1 failure)
    -- failure 1 --
    Expecting actual:
      1650674759819L
    to be greater than:
      1650674760096L
    
    at DefaultClientPollerStatisticsTest.justOccurred(DefaultClientPollerStatisticsTest.java:82)
 ```

Full error:

```
   [INFO] Running org.kiwiproject.dropwizard.poller.metrics.DefaultClientPollerStatisticsTest
    20:46:04.622 [main] WARN org.kiwiproject.dropwizard.poller.metrics.DefaultClientPollerStatistics - Skipping measurement 200 because lock was not obtained in 1 MILLISECONDS
    20:46:04.632 [main] WARN org.kiwiproject.dropwizard.poller.metrics.DefaultClientPollerStatistics - Skipping measurement 100 because lock was not obtained in 1 MILLISECONDS
    [ERROR] Tests run: 16, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.155 s <<< FAILURE! - in org.kiwiproject.dropwizard.poller.metrics.DefaultClientPollerStatisticsTest
    [ERROR] org.kiwiproject.dropwizard.poller.metrics.DefaultClientPollerStatisticsTest.testIncrementCount_SetsLastAttemptTime(SoftAssertions)  Time elapsed: 2.25 s  <<< FAILURE!
    org.assertj.core.error.AssertJMultipleFailuresError: 
    
    Multiple Failures (1 failure)
    -- failure 1 --
    Expecting actual:
      1650674759819L
    to be greater than:
      1650674760096L
    
    at DefaultClientPollerStatisticsTest.justOccurred(DefaultClientPollerStatisticsTest.java:82)
    	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
    	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAfterTestExecutionCallbacks$9(TestMethodTestDescriptor.java:233)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$13(TestMethodTestDescriptor.java:273)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$14(TestMethodTestDescriptor.java:273)
    	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeAllAfterMethodsOrCallbacks(TestMethodTestDescriptor.java:272)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeAfterTestExecutionCallbacks(TestMethodTestDescriptor.java:232)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:137)
    	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
    	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
    	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
    	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
    	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
    	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
    	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
    	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
    	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
    	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
    	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
    	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
    	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
    	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
    	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
    	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
    	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
    	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
    	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
    	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
    	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
    	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:55)
    	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:223)
    	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:175)
    	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:139)
    	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
    	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
    	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
    	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
```

**Comment by sleberknight** [comment]
After looking at the test code, it seems like this is just a timing issue, since we're not mocking the time/date and are using the following method to validate stats:

```java
    private void justOccurred(long timeInMillis, SoftAssertions softly) {
        long now = System.currentTimeMillis();
        softly.assertThat(timeInMillis).isLessThanOrEqualTo(now);
        softly.assertThat(timeInMillis).isGreaterThan(now - TimeUnit.SECONDS.toMillis(1));
    }
```

So, we're expecting `timeInMillis` to be greater than one millisecond before the current time via `System.currentTimeMillis()`.

Relying on this timing in a test is probably not great. In the above example the `timeInMillis` is `1650674759819` while "now - 1 milliscond is `1650674760096`, so `timeInMillis` is actually _less than_ the expected value being compared against by 277 milliseconds.

Without refactoring the test it's probably OK to just increase the number of milliseconds subtracted from `now` to something like 500.

**Comment by sleberknight** [comment]
Oops, totally read the code wrong. We are expecting `timeInMillis` to be greater than one _second_ before the current time, not one millisecond. So obviously increasing the time further isn't a great solution, since 1 second is already an "eternity" in this context. My guess is it might have something to do with the `SoftAssertions`. For whatever reason that can take a while to initialize, so sometimes if a test is the first one run and there is a timing consideration, it can cause test failures if the expected timing tolerance is too tight. This is pure speculation, based on experience with AssertJ's soft assertions.

But, after writing the above, I manually ran only the `testIncrementCount_SetsLastAttemptTime` test 10 times in a row locally in IntelliJ and it failed _every single time_. I also ran the `testIncrementSkipCount_SetsLastSkipTime` test by itself five times and it also failed every time. It also uses soft assertions. The test time reported by IntelliJ was over 2 _seconds_ on every test execution.

I then removed all the soft assertions from these two tests and re-ran them again individually ten times each. They passed every time, with the test execution time ranging from between 200 to almost 500 milliseconds.

So, because these two tests rely on timing (which normally should not be an issue), using AssertJ soft assertions with them is problematic. I suspect the only reason this actually happened is because in one test run, JUnit "decided" to run one of them before any other test, and that caused the lengthy soft assertion "initialization" to occur, which then resulted in a longer time than "one second ago". The solution is therefore to simply remove the soft assertions from these two tests. I don't think losing the soft assertions is that big a deal here from a testing standpoint, since there are only a few (very simple) assertions anyway.
